homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -76,8 +77,15 @@ def icon(self):
         """Icon to use in the frontend, if any."""
         return ICON
 
+    @property
+    def device_state_attributes(self):
+        """Return device specific state attributes."""
+        if self._type == "production":
+            return self._attributes
+
     def update(self):
         """Get the energy production data from the Enphase Envoy."""
         from envoy_reader import EnvoyReader
 
         self._state = getattr(EnvoyReader(self._ip_address), self._type)()
+        self._attributes = EnvoyReader(self._ip_address).inverters_production()
```

Comment:
Two things:
1.  We should not directly assign attributes from the library.  Create a map to copy known elements from the supplied dictionary.  This decouples HA from the library and guards against breaking changes and also ensures that only appropriate items are being returned in state attributes.
2. Shouldn't retrieving the attributes be conditional on whether `self._type` is `production`?

Reply:
@andrewsayre 

1. So I'm not sure how to handle this, the library will return a dictionary of unknown length, with unknown keys. Specifically the length will be the number of solar panel inverters they have, and the keys will be the serial number of each inverter.

2. Done

Reply:
So state attributes are to describe the state of the entity and are things that are likely to change during its life.  These would be things someone would use this to trigger automations or use in scripts and the like.  I don't think serial numbers belong in state attributes.

Reply:
The serial numbers are just a unique identifier for the keys in the dictionary, the values are the things we actually care about, the watts being produced.

Reply:
Why not create separate sensors for the wattage reports? Then we don't need the attributes.

Reply:
@MartinHjelmare Some people might have like 30 of these things, I thought that would be excessive.

Reply:
We prefer discrete sensors for measurements that are useful on their own to describe a device property. The user can decide how they want to display or aggregate the information using the tools available such as lovelace cards, template sensors etc.